Yosemite/Yosemite/Stores/ProductStore.swift

```diff
@@ -1162,25 +1162,25 @@ extension ProductStore {
     /// Updates, inserts, or prunes the provided `storageProduct`'s custom fields using the provided `readOnlyProduct`'s custom fields
     ///
     private func handleProductCustomFields(_ readOnlyProduct: Networking.Product, _ storageProduct: Storage.Product, _ storage: StorageType) {
-        // Upsert the `customFields` from the `readOnlyProduct`
-        readOnlyProduct.customFields.forEach { readOnlyCustomField in
-            if let existingStorageMetaData = storage.loadProductMetaData(siteID: readOnlyProduct.siteID,
-                                                                         productID: storageProduct.productID,
-                                                                         metadataID: readOnlyCustomField.metadataID) {
-                existingStorageMetaData.update(with: readOnlyCustomField)
-            } else {
-                let newStorageMetaData = storage.insertNewObject(ofType: Storage.MetaData.self)
-                newStorageMetaData.update(with: readOnlyCustomField)
-                storageProduct.addToCustomFields(newStorageMetaData)
-            }
-        }
+        // Create a set of metadata IDs from the read-only product for quick lookup
+        let readOnlyMetadataIDs = Set(readOnlyProduct.customFields.map { $0.metadataID })
 
-        // Now, remove any objects that exist in `storageProduct.customFields` but not in `readOnlyProduct.customFields`
+        // Remove any objects that exist in `storageProduct.customFields` regarding a specific product.
         storageProduct.customFields?.forEach { storageCustomField in
-            if readOnlyProduct.customFields.first(where: { $0.metadataID == storageCustomField.metadataID } ) == nil {
-                storageProduct.removeFromCustomFields(storageCustomField)
-                storage.deleteObject(storageCustomField)
-            }
+            storageProduct.removeFromCustomFields(storageCustomField)
+            storage.deleteObject(storageCustomField)
+        }
+
+        // Create `customFields` objects from the `readOnlyProduct`
```

Comment:
Similar to the other comment. I think we should unit tests to validate that it can handle a large number of MetaData items.

Reply:
The test implemented here dd9ca2356ceb1927721114c57c770a18409939ed should work also for Products